Call out the incompatibility of wildcards and L7 permissions (#10725)
hashicorp/consul: website/content/docs/connect/config-entries/service-intentions.mdx

    {
      name: 'Name',
      description:

>         "The name of the destination service for all intentions defined in this config entry. This may be set to the wildcard character (`*`) to match all services that don't otherwise have intentions defined. Wildcard intentions cannot be used when defining L7 [`Permissions`](/docs/connect/config-entries/service-intentions#permissions).",

Reviewer: Would it be worthwhile to codify this in validation logic for the config entry write. That way even if you didn't read the docs there would be an error returned by the API indicating the same thing.

Reviewer: This seems like a breaking change though so it would probably need to wait for the next major release not to mention having some supporting code for stripping out the Permissions for wildcard entries when they are reinserted into memdb.

Reviewer 2: It already behaved this way: https://github.com/hashicorp/consul-enterprise/blob/main/agent/structs/config_entry_intentions.go#L645

Reviewer 2: Paul is just calling out how it already worked better on the docs.

>         "The name of the destination service for all intentions defined in this config entry. This may be set to the wildcard character (`*`) to match all services that don't otherwise have intentions defined. Using a wildcard is incompatible with specifying L7 [`Permissions`](https://www.consul.io/docs/connect/config-entries/service-intentions#permissions) since those can only be enforced for services with the right protocol.",

Reviewer 3: DO you think "Wild card intentions cannot be used when defining L7 permissions" is less ambiguos wording than "is incompatible" ?

>         "The name of the destination service for all intentions defined in this config entry. This may be set to the wildcard character (`*`) to match all services that don't otherwise have intentions defined. Using a wildcard is incompatible with specifying L7 [`Permissions`](https://www.consul.io/docs/connect/config-entries/service-intentions#permissions) since those can only be enforced for services with the right protocol.",

Author: ```suggestion
        "The name of the destination service for all intentions defined in this config entry. This may be set to the wildcard character (`*`) to match all services that don't otherwise have intentions defined. Wildcard intentions cannot be used when defining L7 [`Permissions`](/docs/connect/config-entries/service-intentions#permissions).",
```

      type: `string: "default"`,
      enterprise: true,
      description:

>         "Specifies the namespaces the config entry will apply to. This may be set to the wildcard character (`*`) to match all services in all namespaces that don't otherwise have intentions defined. Using a wildcard is incompatible with specifying L7 [`Permissions`](https://www.consul.io/docs/connect/config-entries/service-intentions#permissions) since those can only be enforced for services with the right protocol.",

Author: ```suggestion
        "Specifies the namespaces the config entry will apply to. This may be set to the wildcard character (`*`) to match all services in all namespaces that don't otherwise have intentions defined. Wildcard intentions cannot be used when defining L7 [`Permissions`](/docs/connect/config-entries/service-intentions#permissions).",
```

          hcl: false,
          type: 'string: <required>',
          description:

>             "The name of the destination service for all intentions defined in this config entry. This may be set to the wildcard character (`*`) to match all services that don't otherwise have intentions defined. Using a wildcard is incompatible with specifying L7 [`permissions`](https://www.consul.io/docs/connect/config-entries/service-intentions#permissions) since those can only be enforced for services with the right protocol.",

Author: ```suggestion
            "The name of the destination service for all intentions defined in this config entry. This may be set to the wildcard character (`*`) to match all services that don't otherwise have intentions defined. Wildcard intentions cannot be used when defining L7 [`Permissions`](/docs/connect/config-entries/service-intentions#permissions).",
```

        },
        {
          name: 'namespace',
          hcl: false,
          enterprise: true,
          type: 'string: <optional>',
          description:

>             "Specifies the namespaces the config entry will apply to. This may be set to the wildcard character (`*`) to match all services in all namespaces that don't otherwise have intentions defined. If not set, the namespace used will depend on the `connectInject.consulNamespaces` configuration. See [ServiceIntentions Special Case (Enterprise)](/docs/k8s/crds#serviceintentions-special-case-enterprise) for more details. Using a wildcard is incompatible with specifying L7 [`permissions`](https://www.consul.io/docs/connect/config-entries/service-intentions#permissions) since those can only be enforced for services with the right protocol.",

Author: ```suggestion
            "Specifies the namespaces the config entry will apply to. This may be set to the wildcard character (`*`) to match all services in all namespaces that don't otherwise have intentions defined. If not set, the namespace used will depend on the `connectInject.consulNamespaces` configuration. See [ServiceIntentions Special Case (Enterprise)](/docs/k8s/crds#serviceintentions-special-case-enterprise) for more details. Wildcard intentions cannot be used when defining L7 [`Permissions`](/docs/connect/config-entries/service-intentions#permissions).",
```

                      provided permissions in this intention will be subject to the default
                      intention behavior is defined by the default [ACL policy](/docs/agent/options#acl_default_policy).<br><br>
                      This should be omitted for an L4 intention as it is mutually exclusive with
                      the \`Action\` field.<br><br>

>                       Setting \`Permissions\` is not valid if a wildcard is used for the \`Name\` or \`Namespace\` since they can only be

Reviewer 4: ```suggestion
                      Setting \`Permissions\` is not valid if a wildcard is used for the \`Name\` or \`Namespace\` because they can only be
```
This is a little grammar nitpick, but a common diction error is using "since" when you mean "because". The "since" preposition relates to time, whereas "because" relates to function. Have some salt with this comment.

Author: Nice tip thanks!

                      applied to services with a compatible protocol.`,
        yaml: `The list of all [additional L7 attributes](#intentionpermission) that extend the intention match criteria.<br><br>
                      Permission precedence is applied top to bottom. For any given request the
                      first permission to match in the list is terminal and stops further
                      evaluation. As with L4 intentions, traffic that fails to match any of the
                      provided permissions in this intention will be subject to the default
                      intention behavior is defined by the default [ACL policy](/docs/agent/options#acl_default_policy).<br><br>
                      This should be omitted for an L4 intention as it is mutually exclusive with
                      the \`action\` field.<br><br>
                      Setting \`permissions\` is not valid if a wildcard is used for the \`spec.destination.name\` or \`spec.destination.namespace\` 

>                       since they can only be applied to services with a compatible protocol.`,

Reviewer 4: ```suggestion
                      because they can only be applied to services with a compatible protocol.`,
```
See previous comment
